File: playground/examples/node/parallel.js
const { task, flow, branch, runDurableWorkflow, InMemoryBackend } = require("sayiir");

const validatePayment = task("validate-payment", (order) => {
  return { ...order, payment: "valid" };
});

const checkInventory = task("check-inventory", (order) => {
  return { ...order, stock: "available" };
});

const finalize = task("finalize", (results) => {
  return `Order complete: ${JSON.stringify(results)}`;
});

const workflow = flow("checkout")
  .fork([branch("payment", validatePayment), branch("inventory", checkInventory)])
  .join("finalize", finalize)
  .build();

const backend = new InMemoryBackend();
const status = runDurableWorkflow(workflow, "run-1", { id: 1, item: "Widget" }, backend);
console.log(status.output);

File: playground/examples/python/parallel.py
from sayiir import task, Flow, run_workflow

@task
def validate_payment(order: dict) -> dict:
    return {**order, "payment": "valid"}

@task
def check_inventory(order: dict) -> dict:
    return {**order, "stock": "available"}

@task
def finalize(results: dict) -> str:
    return f"Order complete: {results}"

workflow = (
    Flow("checkout")
    .fork()
        .branch(validate_payment)
        .branch(check_inventory)
    .join(finalize)
    .build()
)
print(run_workflow(workflow, {"id": 1, "item": "Widget"}))
